### PR TITLE
Save JS "Console..." history across sessions

### PR DESCRIPTION
**Pull request by humbletim**
This PR keeps the history of recent "Console..." commands available across app restarts.

Currently it's possible to use <kbd>up</kbd> and <kbd>down</kbd> to recall and re-execute recent commands at the prompt, but only for the current session.  With this PR the history is automatically saved and restored between sessions to/from a `JSConsole.history.json` file in the app data folder (next to `bookmarks.json`, etc.).

#### Testing
* Launch this PR's Interface.
* Open the "Edit > Console... " prompt.
* Execute some JS commands; eg:
    * `Vec3.print('position', MyAvatar.position);`
    * `MyAvatar.sessionUUID`
    * `print(MyAvatar.getJointIndex('Head'))`
* Exit and restart the application.
* Open the "Edit > Console... " prompt.
* Verify the command history is still available by using <kbd>up</kbd> and <kbd>enter</kbd> to re-execute some commands from the last session.

### CODE REVIEW

**Comment by YuppyPlays** [comment]
Hi @humbletim, does this require QA once built?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1056/HighFidelity-Beta-PR10798-35f27313baeadb9ca3312f1c30ed55d84ac0c50d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1056/HighFidelity-Beta-PR10798-35f27313baeadb9ca3312f1c30ed55d84ac0c50d.exe

**Comment by humbletim** [comment]
@YuppyPlays not quite yet -- currently this is just a suggested change so first needs Development Approved status.

**Comment by YuppyPlays** [comment]
@humbletim Oh right, apologies for my preemptive comment, I shall await until Dev approved and your approval of being ready :)

File: interface/src/ui/JSConsole.cpp [hunk 1]
```diff
@@ -23,28 +23,41 @@
...
 
 const QString GUTTER_PREVIOUS_COMMAND = "<span style=\"color: #57b8bb;\">&lt;</span>";
 const QString GUTTER_ERROR = "<span style=\"color: #d13b22;\">X</span>";
 
 const QString JSConsole::_consoleFileName { "about:console" };
 
+QList<QString> _readLines(const QString& filename) {
+    QFile file(filename);
+    file.open(QFile::ReadOnly);
+    return QTextStream(&file).readAll().split("\r\n");
```
Comment (huffman) [inline interface/src/ui/JSConsole.cpp]: Why `\r\n`?

**Comment by huffman** [comment]
This is really awesome @humbletim !

File: interface/src/ui/JSConsole.cpp (hunk 1 above)
Comment (humbletim) [inline interface/src/ui/JSConsole.cpp]: The idea with using `\r\n` is to keep the history file easy to inspect using built-in tools like Notepad on Windows.

Note that console commands can contain internal new lines -- currently those get normalized to `\n` before appending to history.

Comment (huffman) [inline interface/src/ui/JSConsole.cpp]: Does Qt provide a portable way to do this? `\n` on other platforms?

Comment (humbletim) [inline interface/src/ui/JSConsole.cpp]: ... it turns out `QTextStream` will interpret either variation as a line ending (on any platform) -- so my original strategy simply won't work for multi-line entries.

Based on a suggestion from @thoys I've just committed changes that switch to using platform-agnostic JSON instead.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1718/HighFidelity-Beta-PR10798-9c610f52eca11a94a637a9f408fb6f0c597a0628.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1718/HighFidelity-Beta-PR10798-9c610f52eca11a94a637a9f408fb6f0c597a0628.exe

**Comment by cain-kilgore** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Launch this PR's Interface. **Check**
Open the "Edit > Console... " prompt. **Check**
Execute some JS commands; eg:
Vec3.print('position', MyAvatar.position); 
MyAvatar.sessionUUID 
print(MyAvatar.getJointIndex('Head')) **Each line, typed in separately or pasted all together crashed interface?**
Exit and restart the application. **Cannot continue; no commands appear to be working when typed in**
Open the "Edit > Console... " prompt.
Verify the command history is still available by using up and enter to re-execute some commands from the last session.

**Fail**    Not sure if it's just the example command causing crash? I'm not savvy with the commands but tried entering them as separate lines (for example: "Vec3.print('position', MyAvatar.position); " on it's own and each other line; also tried all in one paste into console, crashed still.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1878/HighFidelity-Beta-PR10798-300b926464e1412be5744316252ba82a1caa8576.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1878/HighFidelity-Beta-PR10798-300b926464e1412be5744316252ba82a1caa8576.exe

**Comment by humbletim** [comment]
I had a bug in my code where [QList::constFirst](http://doc.qt.io/qt-5/qlist.html#constFirst) was being called even when the list was empty (which isn't supported by Qt).

The latest commit specifically checks for this case and the resulting PR builds ought to work as intended now.

**Comment by cain-kilgore** [comment]
@humbletim: Want me to give it a go?

**Comment by humbletim** [comment]
@cain-kilgore Yes that'd be great, thanks!

**Comment by cain-kilgore** [comment]
Cool - testing now!

**Comment by cain-kilgore** [comment]
# Test Passed
Test | Result
-------- | --------
Launch this PR's Interface. | Pass
Open the "Edit > Console... " prompt. | Pass
Execute some JS commands; eg:
Vec3.print('position', MyAvatar.position); | Pass
MyAvatar.sessionUUID | Pass
print(MyAvatar.getJointIndex('Head')) | Pass
Exit and restart the application. | Pass
Open the "Edit > Console... " prompt. | Pass
Verify the command history is still available by using up and enter to re-execute some commands from the last session. | Pass

All looking good @humbletim! :)

###### Intel i5 7600K @ 4Ghz, 16GB DDR4 2133Mhz Corsair Vengeance, Asus Strix GTX 1050Ti 3GB, HTC Vive

**Pull request merged**